src/backend/columnar/columnar_tableam.c

```diff
@@ -1372,21 +1374,145 @@ ColumnarReadRowsIntoIndex(TableScanDesc scan, Relation indexRelation,
+											greatestIndexItemPointer, estate,
+											predicate, validateIndexState);
+
+	table_endscan(scan);
+
+	/* report the last "virtual" block as "done" */
+	ColumnarReportTotalVirtualBlocks(columnarRelation, snapshot,
+									 PROGRESS_SCAN_BLOCKS_DONE);
+
+	ExecDropSingleTupleTableSlot(econtext->ecxt_scantuple);
+	FreeExecutorState(estate);
+	indexInfo->ii_ExpressionsState = NIL;
+	indexInfo->ii_PredicateState = NULL;
+}
+
+
+/*
+ * FindGreatestIndexItemPointer returns ItemPointerData for the index tuple
+ * with greatest tid.
+ * If index is empty, then returns invalid ItemPointerData.
+ */
+static ItemPointerData
+FindGreatestIndexItemPointer(Relation relation, Relation indexRelation, Snapshot snapshot)
+{
+	ItemPointerData greatestIndexItemPointer;
+	ItemPointerSetInvalid(&greatestIndexItemPointer);
+
+	int nkeys = 0;
+	int norderbys = 0;
+	IndexScanDesc indexScan = index_beginscan(relation, indexRelation,
+											  snapshot, nkeys, norderbys);
+	ItemPointer indexItemPointer = NULL;
+	while ((indexItemPointer = index_getnext_tid(indexScan, ForwardScanDirection)))
+	{
+		if (!ItemPointerIsValid(&greatestIndexItemPointer) ||
+			ItemPointerCompare(indexItemPointer, &greatestIndexItemPointer) > 0)
+		{
+			greatestIndexItemPointer = *indexItemPointer;
+		}
+	}
+	index_endscan(indexScan);
+
+	return greatestIndexItemPointer;
+}
+
+
+/*
+ * ColumnarReadRowsWithGreaterTidIntoIndex inserts the tuples that are not in
+ * the index yet by reading the actual relation based on given "scan".
+ */
+static void
+ColumnarReadRowsWithGreaterTidIntoIndex(TableScanDesc scan, Relation indexRelation,
+										IndexInfo *indexInfo,
+										ItemPointerData greatestIndexItemPointer,
+										EState *estate, ExprState *predicate,
+										ValidateIndexState *state)
+{
+	BlockNumber lastReportedBlockNumber = InvalidBlockNumber;
+
+	ExprContext *econtext = GetPerTupleExprContext(estate);
+	TupleTableSlot *slot = econtext->ecxt_scantuple;
+	while (columnar_getnextslot(scan, ForwardScanDirection, slot))
+	{
+		CHECK_FOR_INTERRUPTS();
+
+		BlockNumber currentBlockNumber = ItemPointerGetBlockNumber(&slot->tts_tid);
+		if (lastReportedBlockNumber != currentBlockNumber)
+		{
+			/*
+			 * columnar_getnextslot guarantees that returned tuple will
+			 * always have a greater ItemPointer than the ones we fetched
+			 * before, so we directly use BlockNumber to report our progress.
+			 */
+			Assert(lastReportedBlockNumber == InvalidBlockNumber ||
+				   currentBlockNumber >= lastReportedBlockNumber);
+			pgstat_progress_update_param(PROGRESS_SCAN_BLOCKS_DONE,
+										 currentBlockNumber);
+			lastReportedBlockNumber = currentBlockNumber;
+		}
+
+		MemoryContextReset(econtext->ecxt_per_tuple_memory);
+
+		state->htups += 1;
+
+		if (ItemPointerIsValid(&greatestIndexItemPointer) &&
+			ItemPointerCompare(&slot->tts_tid, &greatestIndexItemPointer) <= 0)
+		{
+			/* tuple is already covered by the index, skip */
+			continue;
```

Comment:
Looks like there is a bug with this approach, we should not assume that tuples having `ItemPointer` less than or equal to the greatest `ItemPointer` that the index has are already indexed. 
Even if we commit the xact at "session 2", we didn't yet index the tuples inserted here.
This is because, we indexed the tuples that "session 4" inserted and greatest `ItemPointer` that the index has becomes larger than the ones insert in "session 2", so we just skip them:

```sql
--
-- session 1 --
--
-- break at columnar_index_build_range_scan, also break at validate scan
DROP TABLE IF EXISTS columnar_table;
CREATE TABLE columnar_table (a INT, b int) USING columnar;
INSERT INTO columnar_table SELECT i FROM generate_series(1, 25000) i;
-- set breakpoint at columnar_index_build_range_scan
CREATE INDEX CONCURRENTLY ON columnar_table (a);
-- continue with session 2 & 3 & 4 after hitting to breakpoint at columnar_index_build_range_scan

--
-- session 2
--
BEGIN;
INSERT INTO columnar_table SELECT i FROM generate_series(26001, 27000) i;

--
-- session 3
--
BEGIN;
INSERT INTO columnar_table SELECT i FROM generate_series(28001, 29000) i;

--
-- session 4
--
INSERT INTO columnar_table SELECT i FROM generate_series(30001, 31000) i;

--
-- session 1
--
-- blocks until xacts at session 2 & 3 are committed or aborted

--
-- session 2
--
COMMIT;

--
-- session 3
--
ROLLBACK;

--
-- session 1
--
-- CREATE INDEX CONCURRENTLY finishes now

-- force columnar custom scan
SET enable_seqscan to off; SET columnar.enable_custom_scan to on; SET enable_indexscan to off;
SELECT a FROM columnar_table WHERE a = 26501;
┌───────┐
│   a   │
├───────┤
│ 26501 │
└───────┘
(1 row)

-- force index scan
SET enable_seqscan to off; SET columnar.enable_custom_scan to off; SET enable_indexscan to on;

SELECT a FROM columnar_table WHERE a = 26501;
┌───┐
│ a │
├───┤
└───┘
(0 rows)
```

Reply:
We should follow the very same logic from heapAM:

https://github.com/postgres/postgres/blob/dbab0c07e5ba1f19a991da2d72972a8fe9a41bda/src/backend/access/heap/heapam_handler.c#L1860-L1907

Reply:
Looks like I fixed that: 5611438

Reply:
I found one more bug on Friday:

```sql
DROP TABLE IF EXISTS columnar_table CASCADE;
CREATE TABLE columnar_table (a INT, b INT) USING columnar;

-- s1
BEGIN;
INSERT INTO columnar_table SELECT i, 1000*i FROM generate_series(1, 10) i;

-- s2
CREATE INDEX CONCURRENTLY conc_s2_idx ON columnar_table(a);

-- s5
BEGIN;
INSERT INTO columnar_table SELECT i, i*1000 FROM generate_series(11, 20) i;

-- s3
BEGIN;
INSERT INTO columnar_table SELECT i, i*1000 FROM generate_series(21, 30) i;

-- s1
COMMIT;

-- s4 -- bu iki kere geliyo
INSERT INTO columnar_table SELECT i, i*1000 FROM generate_series(31, 40) i;

-- s5
COMMIT;

-- s3
ROLLBACK;

-- s1
-- force index scan
SET enable_seqscan TO OFF; SET columnar.enable_custom_scan TO OFF; SET enable_indexscan TO ON;
-- Shows "35" two times. Indeed, tuples inserted by "s4" are duplicated
SELECT a FROM columnar_table WHERE a=35;
```

Reply:
Looks like fixed that too 83d995b973475264dd587314d15d316e2eb8210f.